assets/app/scripts/directives/oscFileInput.js

```diff
@@ -10,15 +10,121 @@ angular.module('openshiftConsole')
         disabled: "=ngDisabled",
         // Show the file contents below the file input.
         showValues: "=",
-        helpText: "@?"
+        helpText: "@?",
+        dropZoneId: "@?",
+        dragging: "=",
       },
       templateUrl: 'views/directives/osc-file-input.html',
       link: function(scope, element){
         scope.helpID = _.uniqueId('help-');
         scope.supportsFileUpload = (window.File && window.FileReader && window.FileList && window.Blob);
         scope.uploadError = false;
-        $(element).change(function(){
-          var file = $('input[type=file]', this)[0].files[0];
+
+        var droppableElement = (scope.dropZoneId) ? $('#' + scope.dropZoneId) : element,
+        dropZoneName = scope.dropZoneId + "-drag-and-drop-zone",
+        dropZoneSelectorName = "#" + dropZoneName,
+        highlightDropZone = false,
+        showDropZone = false,
+        addDropZone = false;
+
+        scope.$watch('disabled', function() {
+          if (scope.disabled) {
+            // Remove listeners for dropZone drag & drop actions
+            removeDropZoneListeners();
+            removeDropZoneElements();
+            removeDocumentListeners();
+          } else {
+            // Add drop, dragover, dragleave listeners to prevent redirecting upon dropping file outside the dropZone.            
+            $(document).on('dragover', function(e) {
+              preventAndStop(e);
+
+              showDropZone = true;
+              // Prevent adding additional elements and registering listeners by setting 'addDropZone' to 'true'
+              if (!addDropZone) {
+                addDropZone = true;
+                addZoneToElement();
+                addDropZoneListeners();
+              }
+            });
+
+            $(document).on('dragleave', function(e) {
+              preventAndStop(e);
+              showDropZone = false;
+              _.delay(function(){
+                if( !showDropZone ){
+                  addDropZone = false;
+                  removeDropZoneElements();
+                  removeDropZoneListeners();
+                }
+              }, 200 );
+            });
+          }
+        }, true);
+
+        // Add handler for dropping file out of the dragZone only once
+        if (_.isUndefined($._data( $(document)[0], "events").drop)) {
+          $(document).on('drop', function(e) {
+            preventAndStop(e);              
+            $('.drag-and-drop-zone').trigger('reset');
```

Comment:
Everywhere we reference `.drag-and-drop-zone`, we should probably use

```
$(element).find('.drag-and-drop-zone') ...
```

so that we only target children of this specific file input. Otherwise, you will update all of them on the page if there are other file inputs.